src/libraries/System.Linq/src/System/Linq/MaxMin.cs

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        private interface IMinMaxCalc<T> where T : struct, IBinaryInteger<T>
+        {
+            public static abstract bool Compare(T left, T right);
+            public static abstract Vector128<T> Compare(Vector128<T> left, Vector128<T> right);
+            public static abstract Vector256<T> Compare(Vector256<T> left, Vector256<T> right);
+        }
+
+        private static T MinMaxInteger<T, TMinMax>(this IEnumerable<T> source)
+            where T : struct, IBinaryInteger<T>
+            where TMinMax : IMinMaxCalc<T>
+        {
+            T value;
+
+            if (source.TryGetSpan(out ReadOnlySpan<T> span))
+            {
+                if (span.IsEmpty)
+                {
+                    ThrowHelper.ThrowNoElementsException();
+                }
+
+                if (!Vector128.IsHardwareAccelerated || span.Length < Vector128<T>.Count)
+                {
+                    value = span[0];
+                    for (int i = 1; i < span.Length; i++)
+                    {
+                        if (TMinMax.Compare(span[i], value))
+                        {
+                            value = span[i];
+                        }
+                    }
+                }
+                else if (!Vector256.IsHardwareAccelerated || span.Length < Vector256<T>.Count)
+                {
+                    ref T current = ref MemoryMarshal.GetReference(span);
+                    ref T lastVectorStart = ref Unsafe.Add(ref current, span.Length - Vector128<T>.Count);
+
+                    Vector128<T> best = Vector128.LoadUnsafe(ref current);
+                    current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+
+                    while (Unsafe.IsAddressLessThan(ref current, ref lastVectorStart))
+                    {
+                        best = TMinMax.Compare(best, Vector128.LoadUnsafe(ref current));
+                        current = ref Unsafe.Add(ref current, Vector128<T>.Count);
+                    }
+                    best = TMinMax.Compare(best, Vector128.LoadUnsafe(ref lastVectorStart));
+
+                    value = best[0];
+                    for (int i = 1; i < Vector128<T>.Count; i++)
+                    {
+                        if (TMinMax.Compare(best[i], value))
+                        {
+                            value = best[i];
+                        }
+                    }
```

Comment:
I've probably missed it but where is the code checking for not aligned part?

Reply:
What do you mean?

Reply:
say you got 11 elements and you check by 4:
`[1, 3, 2, 7,    2, 5, 6, 1,   9, 2, 6]`

where does check for `[9, 2, 6]` happens? it seems the IsAddressLessThan will skip them

Reply:
(or go out of bands)

Reply:
> where does check for [9, 2, 6] happens? it seems the IsAddressLessThan will skip them

Here:
https://github.com/dotnet/runtime/pull/76144/files#diff-1442451023271f0880ec601e6fb9b85c4ca55140579fe82c6c212c8c9d85c04aR58

Reply:
but how does it know how many elements it's safe to read?

Reply:
It reads a vector's worth from the end:
https://github.com/dotnet/runtime/pull/76144/files#diff-1442451023271f0880ec601e6fb9b85c4ca55140579fe82c6c212c8c9d85c04aR48

This may overlap with previously processed elements.  That's fine, because the operation is idempotent.

Reply:
that makes sense now, thanks for explanation!